File: leetcode/#260.py
# question --> https://leetcode.com/problems/single-number-iii/

class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        lookup = set()

        for number in nums: 
            if number in lookup: 
                lookup.remove(number)
            elif number not in lookup: 
                lookup.add(number)

        return sorted(lookup) 
        
        # TC: O(N); SC: O(N)

# Optimised Approach - 01

class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        xor = 0
        for num in nums: 
            xor ^= num
            
        firstSetBit = 0 
                
        for firstSetBit in range(32):
            if (1 << firstSetBit) & xor: 
                break
                
        num1 = 0
        for num in nums: 
            if (1 << firstSetBit) & num: 
                num1 ^= num
                
        return [min(num1, num1^xor), max(num1, num1^xor)]
        
        # TC: O(N); SC: O(1)
        
# Optimised Approach - 02

class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        xor = 0
        for num in nums: 
            xor ^= num
            
        firstSetBit = (xor & (xor - 1)) ^ xor
                
        num1 = 0
        for num in nums: 
            if firstSetBit & num: 
                num1 ^= num
                
        return [min(num1, num1^xor), max(num1, num1^xor)]
        
        # TC: O(N); SC: O(1)
        
# end 
        
         

File: leetcode/#260.js
// question --> https://leetcode.com/problems/single-number-iii/

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    let lookup = new Set();
    
    for (let index = 0; index < nums.length; index++) {
        if (lookup.has(nums[index])) {
            lookup.delete(nums[index]); 
        }
        else {
            lookup.add(nums[index]); 
        }
    }
    
    lookup = Array.from(lookup);
    return lookup.sort(); 
};

// TC: O(N); SC: O(N)

// Optimised Approach - 01

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    let xor = 0; 
    for (let index = 0; index < nums.length; index++) {
        xor ^= nums[index];
    } 
    let firstSetBit = 0; 
    for (firstSetBit; firstSetBit < 32; firstSetBit++) {
        if ((1 << firstSetBit) & xor) {
            break; 
        }
    }
    let num1 = 0; 
    for (let index = 0; index < nums.length; index++) {
        if ((1 << firstSetBit) & nums[index]) {
            num1 ^= nums[index]; 
        }
    }
    let ans = [Math.min(num1, num1^xor), Math.max(num1, num1^xor)];
    return ans
};

// TC: O(N); SC: O(1)

// Optimal Approach - 02

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    let xor = 0; 
    for (let index = 0; index < nums.length; index++) {
        xor ^= nums[index];
    } 
    let firstSetBit = (xor & (xor - 1)) ^ xor;
    let num1 = 0; 
    for (let index = 0; index < nums.length; index++) {
        if (firstSetBit & nums[index]) {
            num1 ^= nums[index]; 
        }
    }
    let ans = [Math.min(num1, num1^xor), Math.max(num1, num1^xor)];
    return ans
};

// TC: O(N); SC: O(1)

// end 

